File: nakamura-makoto/BaseVtkLegacyConverter.java
package vtkConverter;

import java.io.IOException;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.DoubleStream;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public abstract class BaseVtkLegacyConverter{
    public void execute(Path inputPath,Path outputPath) throws IOException {

    	Object3D output3DObject = this.readFile(inputPath);
    	this.writeFile(output3DObject, outputPath);
    }

    abstract protected Object3D readFile(Path inputPath) throws IOException;

    private void writeFile(Object3D output3DObject, Path outputPath) throws IOException {
        if (!Files.exists(outputPath)) {
            Files.createFile(outputPath);
        }
        //write header
        List<String> header = Arrays.asList(
        		"# vtk DataFile Version 4.2",
        		"vtk output",
        		"ASCII",
        		"DATASET UNSTRUCTURED_GRID");

        //write points section
        List<String> pointsSection = new ArrayList<>();

        int numPoints = output3DObject.getPoints().size();
        pointsSection.add("POINTS " + String.valueOf(numPoints) + " float");

        for(int i = 0; i < numPoints; i += 3) {
            double[] xyz = output3DObject.getPoints().get(i);
            double[] xyz2 = output3DObject.getPoints().get(i + 1);
            double[] xyz3 = output3DObject.getPoints().get(i + 2);
            String xyzStr = Stream.of(xyz,xyz2,xyz3).flatMapToDouble(e-> Optional.ofNullable(e)
          		  .map(DoubleStream::of).orElse(DoubleStream.empty())
          		  ).boxed()
          		.map(e->{
          			BigDecimal bd = new BigDecimal(String.valueOf(e));
          			int scale = 6 - (bd.precision() - bd.scale());
          			bd = bd.setScale(scale, RoundingMode.HALF_UP);
          			return bd.toPlainString();
          		}).collect(Collectors.joining(" "));

          pointsSection.add(xyzStr + " ");
        }

        // write cells section
        List<String> cellsSection = new ArrayList<>();

        int numCells = output3DObject.getCells().size();
        long numCellsAttributes = output3DObject.getCells().values().stream()
        		.flatMapToInt(e-> Optional.ofNullable(e)
            		  .map(IntStream::of).orElse(IntStream.empty())
            		  ).boxed().count();
        numCellsAttributes += numCells;
        System.out.println(numCellsAttributes);
        cellsSection.add("CELLS " + String.valueOf(numCells)
        		+ " " + String.valueOf(numCellsAttributes) + " ");

        for(int i = 0;i < numCells;i++) {
        	int[] cell = output3DObject.getCells().get(i);
        	String cellStr = IntStream.of(cell).boxed().map(s->s.toString())
        			.collect(Collectors.joining(" "));

            String cellInfoStr = String.valueOf(cell.length) + " " + cellStr + " ";
            cellsSection.add(cellInfoStr);
        }
        cellsSection.add("");

        // write cellTypes section
        List<String> cellTypesSection = new ArrayList<>();
        cellTypesSection.add("CELL_TYPES " + String.valueOf(numCells));
        for(int i = 0;i < numCells; i++) {
        	String cellTypeStr = output3DObject.getCellTypes().get(i);
        	int cellType = this.cellTypeFromStr2Vtk(cellTypeStr);
        	cellTypesSection.add(String.valueOf(cellType));
        }

        cellTypesSection.add("");

        List<String> allSection = new ArrayList<>() {
        	{
        		addAll(header);
        		addAll(pointsSection);
        		addAll(cellsSection);
        		addAll(cellTypesSection);
        	}
        };
        Files.write(outputPath, allSection, Charset.forName("UTF-8"), StandardOpenOption.WRITE);
    }

    private int cellTypeFromStr2Vtk(String cellTypeStr) {
        int retCellType = -1;

        if (cellTypeStr.equals("tri")) {
        	retCellType = 5;
        }else if(cellTypeStr.equals("tetra")) {
        	retCellType = 10;
        }
        return retCellType;
    }
}

File: kozuka-toma/base_vtk_legacy_converter.py

class Object3D(object):
    def __init__(self):
        self.points = list(list())  # [[x0, y0, z0], [x1, y1, z1], ...]
        self.cells = list(list())  # [[pid00, pid01, ...], [pid10, pid11, ..]]
        self.cell_types = list()  # [cell_type0, cell_type1, ...]


class BaseVtkLegacyConverter(object):
    def __init__(self):
        self.object_3d = Object3D()

    def Execute(self, in_file_name, out_file_name):
        self.ReadFile(in_file_name)
        print("Done parsing.")
        self.WriteFile(out_file_name)
        print("Done writing.")

    def ReadFile(self, in_file_name):
        with open(in_file_name, mode="r") as file_obj:
            line_str_list = file_obj.readlines()
        return line_str_list


    def WriteFile(self, out_file_name):
        with open(out_file_name, mode="w") as file_obj:
            # Write header. ---
            file_obj.write("# vtk DataFile Version 4.2\n")
            file_obj.write("vtk output\n")
            file_obj.write("ASCII\n")
            file_obj.write("DATASET UNSTRUCTURED_GRID\n")
            # --- Write header.

            # Write points section. ---
            num_points = len(self.object_3d.points)
            file_obj.write("POINTS " + str(num_points) + " float\n")
            for point_id in range(num_points):
                xyz = self.object_3d.points[point_id]
                xyz_str = [str(elem) for elem in xyz]
                xyz_str = " ".join(xyz_str) + " "
                file_obj.write(xyz_str)
                if (point_id+1) % 3 == 0:
                    file_obj.write("\n")
            # --- Write points section.

            file_obj.write("\n")

            # Write cells section. ---
            num_cells = len(self.object_3d.cells)
            num_cells_attributes = sum([len(cell)
                                        for cell in self.object_3d.cells])
            num_cells_attributes += num_cells
            file_obj.write("CELLS " + str(num_cells) + " " +
                           str(num_cells_attributes) + " \n")
            for cell_id in range(num_cells):
                cell = self.object_3d.cells[cell_id]
                cell_str = [str(point_id) for point_id in cell]
                cell_str = " ".join(cell_str)

                cell_info_str = str(len(cell)) + " " + cell_str + "\n"
                file_obj.write(cell_info_str)
            # --- Write cells section.

            file_obj.write("\n")

            # Write cell_types section. ---
            file_obj.write("CELL_TYPES " + str(num_cells) + "\n")
            for cell_id in range(num_cells):
                cell_type_str = self.object_3d.cell_types[cell_id]
                cell_type = self.CellTypeFromStr2Vtk(cell_type_str)
                file_obj.write(str(cell_type))
                file_obj.write("\n")
            # --- Write cells section.

    def CellTypeFromStr2Vtk(self, cell_type_str):
        ret_cell_type = -1
        if cell_type_str == "tri":
            ret_cell_type = 5
        elif cell_type_str == "tetra":
            ret_cell_type = 10
        return ret_cell_type
